test/winter_onboarding_2021/fleet_management_service/handlers/user_test.clj

```diff
@@ -10,63 +14,164 @@
+                  :g-recaptcha-response "abcd123"}
+            response (handler/create-user {:form-params user})]
+        (is (= (get-in response [:flash :message])
+               (str "Token expired.")))))
+    (testing "Should return token usage limit reached"
+      (let [admin (factories/admin)
+            _ (org-handler/create {:user admin
+                                   :form-params {:name "First Organization"}})
+            invite (factories/invite-driver {:invites/created-by (:users/id admin)
+                                             :invites/usage-limit 1})
+            host (first (user-model/find-by-keys {:users/email (:users/email admin)}))
+            user1 {:name "Pomona Sprout"
+                   :email "p.sprout@hogwarts.edu"
+                   :role "admin"
+                   :password "pulses"
+                   :token (:invites/token invite)
+                   :g-recaptcha-response "abcd123"}
+            user2 {:name "Remus Lupin"
+                   :email "r.lupin@hogwarts.edu"
+                   :role "admin"
+                   :password "moony"
+                   :token (:invites/token invite)
+                   :g-recaptcha-response "abcd123"}
+            response1 (handler/create-user {:form-params user1})
+            response2 (handler/create-user {:form-params user2})]
+        (is (= (get-in response1 [:flash :message])
+               (str "User Pomona Sprout created successfully in organisation " (:users/org-id host))))
+        (is (= (get-in response2 [:flash :message])
+               (str "Token usage limit reached. Please use a new token")))))
+    (testing "Should return 'Invite Token not found' and NOT create user"
+      (let [user {:name "Albus McGonagall"
+                  :email "albus.minerva@hogwarts.edu"
+                  :role "admin"
+                  :password "albus"
+                  :token "ABCDEF"
+                  :g-recaptcha-response "abcd123"}
+            response (handler/create-user {:form-params user})
+            new-user (first (user-model/find-by-keys {:users/email (:email user)}))]
+        (is (= (get-in response [:flash :message])
+               "Invite Token not found"))
+        (is (= new-user
+               nil))))))
 
 (deftest user-login
   (testing "Correct login credentials, should redirect to user dashboard"
-    (with-redefs [utils/uuid (fn [] (java.util.UUID/fromString "9088992d-d0f4-4207-9b95-c934ad071c32"))]
+    (with-redefs [utils/uuid (fn [] (java.util.UUID/fromString "9088992d-d0f4-4207-9b95-c934ad071c32"))
+                  client/post (constantly {:body (json/json-str {"success" true})})]
       (let [user {:name "Severus Snape"
                   :email "s.snape@hogwarts.edu"
                   :role "admin"
-                  :password "lily"}
+                  :password "lily"
+                  :g-recaptcha-response "abcd123"}
             _ (handler/create-user {:form-params user})
```

Comment:
In this `with-redefs`, it returns a success message regardless of the `:g-captcha-response` in the `user` hashmap. Could we check that `:g-captcha-response` is actually `"abcd123"` inside that redefinition?

Why? Because I couldnt figure out where `client/post` is being called in the code & then it dawned on me that we are doing it inside the handler. Doing the above might help in readability?

Reply:
Since we are doing this redefinition in other tests as well, could we just stub this `client/post` function and use it everywhere redefinition is needed?

Reply:
Can you elaborate more?